CacheManager.StackExchange.Redis.SelfCleaning.Tests/SelfCleaningRedisIntegrationTests.cs

```diff
@@ -0,0 +1,247 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+using CacheManager.Core;
+using CacheManager.Core.Internal;
+using CacheManager.StackExchange.Redis.SelfCleaning.Core;
+using CacheManager.StackExchange.Redis.SelfCleaning.Timers;
+using Moq;
+using NUnit.Framework;
+using StackExchange.Redis;
+
+namespace CacheManager.StackExchange.Redis.SelfCleaning.Tests
+{
+    [TestFixture]
+    public class SelfCleaningRedisIntegrationTests
+    {
+        private const double TIME_TO_LIVE_MILLISECONDS_DIFFERENCE_THRESHOLD = 100;
+
+        private Mock<IServer> _serverMock;
+        private Mock<IDatabase> _databaseMock;
+        private Mock<IConnectionMultiplexer> _connectionMock;
+
+        private IDictionary<RedisKey, (RedisValue[] Values, DateTime InsertionTime)> _fauxDatabase;
+
+        private TimeSpan _timeToLive;
+        private ICacheManager<DummyModel> _cache;
+
+        private ICollection<(CacheItemRemovedEventArgs Args, DateTime Time)> _onRemoveByHandleInvocations;
+
+        [OneTimeSetUp]
+        public void Setup()
+        {
+            _serverMock = new Mock<IServer>();
+            _databaseMock = new Mock<IDatabase>();
+            _connectionMock = new Mock<IConnectionMultiplexer>();
+
+            _fauxDatabase = new ConcurrentDictionary<RedisKey, (RedisValue[], DateTime)>();
+
+            SetupServerMock();
+            SetupDatabaseMock();
+            SetupConnectionMock();
+            SetupCache();
+        }
+
+        #region Setup Methods
+
+        private void SetupServerMock()
+        {
+            _serverMock.SetupGet(server => server.IsConnected).Returns(true);
+            _serverMock.SetupGet(server => server.Features).Returns(new RedisFeatures(Version.Parse("3.0.504")));
+            _serverMock
+                .Setup(server => server.Keys(It.IsAny<int>(), It.IsAny<RedisValue>(), It.IsAny<int>(), It.IsAny<long>(),
+                    It.IsAny<int>(), It.IsAny<CommandFlags>()))
+                .Returns<int, RedisValue, int, long, int, CommandFlags>(
+                    (db, pattern, pageSize, cursor, pageOffset, flags) => _fauxDatabase.Keys);
+        }
+
+        private void SetupDatabaseMock()
+        {
+            _databaseMock.SetupGet(database => database.Database).Returns(0);
+            _databaseMock.Setup(database => database.KeyIdleTime(It.IsAny<RedisKey>(), It.IsAny<CommandFlags>()))
+                .Returns<RedisKey, CommandFlags>((key, flags) => DateTime.Now - _fauxDatabase[key].InsertionTime);
+
+            void Put(byte[] hash, RedisKey[] keys, RedisValue[] values, CommandFlags flags)
+            {
+                RedisValue[] newValues = {values[0], values[2], values[3], values[4], values[1], values[5], true};
+                (RedisValue[], DateTime) tuple = (newValues, DateTime.Now);
+
+                RedisKey key = keys[0];
```

Comment:
If you force the keys to have only 1 key inside, why not adding that to the setup of ScriptEvaluate?
`It.Is<RedisKey[]>(keys => keys.Length == 1)`

Reply:
Technically your'e correct, but I don't see how this matters since it's happens in both `Put` and `Get`